newshomepages/hyperlinks.py

```diff
@@ -52,25 +51,16 @@ def _get_links(context: BrowserContext, data: dict, timeout: int = 180):
 
     # Go to the page
     page.goto(data["url"], timeout=timeout * 1000)
-
-    # Pull the html
-    html = page.content()
-
-    # Parse out all the links
-    soup = BeautifulSoup(html, "html5lib")
-    link_list = soup.find_all("a")
-
-    # Parse out the data we want to keep
-    data_list = []
-    for link in link_list:
-        try:
-            d = dict(text=link.text, url=link["href"])
-        except KeyError:
-            # If no href, skip it
-            continue
-
-        # Add to big list
-        data_list.append(d)
+    # Parse out the data we want to keep in JavaScript
+    link_list = page.evaluate(
+        """
+                              Array.from(
+                                document.getElementsByTagName("a"))
+                                    .map((a) => [a, a.getBoundingClientRect()])
+                                    .map(([a, rect]) => ({"text": a.text, "url": a.href, "x": rect.x, "y": rect.y, "w": rect.width, "h": rect.height })
```

Comment:
My main concern about the x/y/w/h object is that it's not explicitly clear that the x and y refer to the coordinates of the upper left corner.

One "canonical" standard we could lean on for our data structure is the [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) method itself. While not perfect, it's pretty easy to justify, since that's the method we're using to draw out the links.

It includes the classic browser properties of top/right/bottom/left, with top serving as an alias to y and left serving as an alias to left. How would you feel about substituting those four properties for the the x and y we have now?

Reply:
Sure! Will do.